rest/api_test.go

```diff
@@ -5662,3 +5662,341 @@ func TestUserXattrSyncFn(t *testing.T) {
 	require.NoError(t, err)
 	assert.Equal(t, []string{"channel"}, doc.SyncData.Channels.KeySet())
 }
+
+func TestUserXattrAutoImport(t *testing.T) {
+	if base.UnitTestUrlIsWalrus() {
+		t.Skip("This test only works against Couchbase Server")
+	}
+
+	defer base.SetUpTestLogging(base.LevelDebug, base.KeyAll)()
+
+	docKey := t.Name()
+	xattrKey := "myXattr"
+	channelName := "testChan"
+
+	// Sync function to set channel access to whatever xattr is
+	rt := NewRestTester(t, &RestTesterConfig{
+		DatabaseConfig: &DbConfig{
+			AutoImport:   true,
+			UserXattrKey: xattrKey,
+		},
+		SyncFn: `
+			function (doc, oldDoc, meta){
+				if (meta.xattrs.myXattr !== undefined){
+					channel(meta.xattrs.myXattr);
+				}
+			}`,
+	})
+
+	defer rt.Close()
+
+	gocbBucket, ok := base.AsGoCBBucket(rt.Bucket())
+	if !ok {
+		t.Skip("Test requires Couchbase Bucket")
+	}
+
+	// Add doc
+	resp := rt.SendAdminRequest("PUT", "/db/"+docKey, "{}")
+	assertStatus(t, resp, http.StatusCreated)
+
+	// Add xattr to doc
+	_, err := base.WriteXattr(gocbBucket, docKey, xattrKey, channelName)
+	assert.NoError(t, err)
+
+	// Wait for doc to be imported
+	err = rt.WaitForCondition(func() bool {
+		return rt.GetDatabase().DbStats.SharedBucketImport().ImportCount.Value() == 1
+	})
+	assert.NoError(t, err)
+
+	// Ensure sync function has ran twice (once for PUT and once for xattr addition)
+	assert.Equal(t, int64(2), rt.GetDatabase().DbStats.CBLReplicationPush().SyncFunctionCount.Value())
+
+	// Get Xattr and ensure channel value set correctly
+	var syncData db.SyncData
+	_, err = gocbBucket.GetXattr(docKey, base.SyncXattrName, &syncData)
+	assert.NoError(t, err)
+
+	assert.Equal(t, []string{channelName}, syncData.Channels.KeySet())
+
+	// Update xattr again but same value and ensure it isn't imported again (crc32 hash should match)
+	_, err = base.WriteXattr(gocbBucket, docKey, xattrKey, channelName)
+	assert.NoError(t, err)
+
+	var syncData2 db.SyncData
+	_, err = gocbBucket.GetXattr(docKey, base.SyncXattrName, &syncData2)
+	assert.NoError(t, err)
+
+	assert.Equal(t, syncData.Crc32c, syncData2.Crc32c)
+	assert.Equal(t, syncData.Crc32cUserXattr, syncData2.Crc32cUserXattr)
+	assert.Equal(t, int64(2), rt.GetDatabase().DbStats.CBLReplicationPush().SyncFunctionCount.Value())
```

Comment:
Why not use the import count stat here, as you do below?

Reply:
Import shouldn't occur in this case but I guess I can check that it remains the same

Reply:
What I meant was that although we expect SyncFunctionCount to equal ImportCount in this particular usage, it might make more sense to just use ImportCount and omit SyncFunctionCount.

Reply:
I can just remove if we don't think its appropriate but I think we might as well leave it in. I can't think of any reason why the sync function count wouldn't be equal to the import count here but its not impossible that something would break here in the future.